contributors/design-proposals/kubeadm-upgrades.md

```diff
@@ -0,0 +1,350 @@
+
+#### Static Pod-hosted control plane
+
+When the upgrade functionality is introduced in v1.8; the default way to set up the control plane is by using Static Pods.
+
+When the control plane is hosted using Static Pods, the actual upgrade flow looks like this:
+ - Gets the Mirror Pod manifests from the API Server (`kubectl -n kube-system get pod ${component}-${nodename} -ojson`), and sha256-hashes the JSON content.
+ - Writes new Static Pod manifests to a **temporary directory** (with the prefix `/etc/kubernetes/tmp/kubeadm-upgraded-manifests*`) based on the new configuration object built internally earlier.
+ - Creates a **backup directory** with the prefix `/etc/kubernetes/tmp/kubeadm-backup-manifests*`.
+   - Q: Why `/etc/kubernetes/tmp`?
+   - A: Possibly not very likely, but we concluded that there may be an attack area for computers where `/tmp` is shared and writable by all users.
+     We wouldn't want anyone to mock with the new Static Pod manifests being applied to the clusters. Hence we chose `/etc/kubernetes/tmp`, which is
+     root-only owned.
+ - In a loop for all control plane components:
+   - Moves the **current manifest** (in `/etc/kubernetes/manifests`) to the **backup directory** (`/etc/kubernetes/tmp/kubeadm-backup-manifests*/`)
+   - Moves the **upgraded manifest** (in `/etc/kubernetes/tmp/kubeadm-upgraded-manifests*`) to the **Static Pod manifest folder** (`/etc/kubernetes/manifests`)
+     - => kubelet notices that the Static Pod on disk changed, kills the old Pod and creates the new one.
+   - kubeadm waits for the API server to come up, executes `kubectl -n kube-system get pod ${component}-${nodename} -ojson`, sha256-hashes the response,
+     and compares it to the pre-upgrade hash. If the hash has changed, kubeadm proceeds with the process.
+     - _Race condition warning_: If the Pod manifest hashing wouldn't be performed, there is a very common case where kubeadm upgrades the Static Pod manifest
+       on disk, but proceeds too quickly (before the kubelet has actually restarted the Pod), and falsely checks that the old component is running.
+   - kubeadm waits for the Static Pod's Mirror Pod to register itself with the API Server; just as an extra security measure
+ - In order to not leak directories on the host, kubeadm deletes the **backup** and **temporary directory**.
+   - If there are user requests, we might make it possible to opt-out of the "always delete old manifests"-policy.
+
+##### Rollbacks
+
+In case any of these operations fail, kubeadm will rollback the earlier manifests to the previous state. Hence the backup directory is important.
+For instance, if the scheduler doesn't come up cleanly; kubeadm will rollback the previously (successfully upgraded) API server, controller manager
+manifests as well as the scheduler manifest.
+
+#### Self-hosted control plane
```

Comment:
it looks like this section isn't finished so i'm going to stop reviewing here.

Reply:
Yeah, thanks for the review so far!